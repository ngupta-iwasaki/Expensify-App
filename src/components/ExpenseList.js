import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListitem";
import selectExpenses from "../selectors/expenses";
import ExpenseListFilter from "./ExpenseListFilter";

export const ExpenseList = (props) => (
    <div>
        {props.expenses.length === 0 ?
            <p>There are no expenses</p> : 
            props.expenses.map(exp => (
                <ExpenseListItem key={exp.id}
                                    {...exp} />))}
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
};

export default connect(mapStateToProps)(ExpenseList);