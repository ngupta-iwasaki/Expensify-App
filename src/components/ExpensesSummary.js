import React from "react";
import {connect} from "react-redux";
import numeral from "numeral";
import getExpensesTotal from "../selectors/expenses-total";
import getVisibleExpenses from "../selectors/expenses";

export const ExpensesSummary = (props) => (
    <h1>
        Viewing {props.expenseCount}
         {props.expenseCount === 1 ? " expense " : " expenses "}
         totalling {numeral(props.expensesTotal / 100).format("$0,0.00")} 
    </h1>
);

const mapStateToProps = (state) => {
    const visibleExpenses = getVisibleExpenses(state.expenses, state.filters);
    
    return {
        expenseCount: visibleExpenses.length,
        expensesTotal: getExpensesTotal(visibleExpenses)
    };
};

export default connect(mapStateToProps)(ExpensesSummary);