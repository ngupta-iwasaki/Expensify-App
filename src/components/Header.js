import React from 'react';
import {connect} from "react-redux";
import {NavLink} from "react-router-dom";
import {startLogout} from "../actions/auth.js";

export const Header = (props) => (
    <header>
        <h1>Expensify</h1>
        <NavLink to="/dashboard" exact={true} activeClassName="isActive">Dashboard</NavLink>
        <NavLink to="/create" activeClassName="isActive">Create Expense</NavLink>
        <button onClick={props.startLogout}>Log Out</button>
    </header>
);

const mapDispatchToProps = (dispatch) => ({
    startLogout: () => dispatch(startLogout())
});

export default connect(undefined, mapDispatchToProps)(Header);