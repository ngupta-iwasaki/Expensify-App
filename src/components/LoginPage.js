import React from "react";
import {connect} from "react-redux";
import {startLogin} from "../actions/auth.js";

export const LoginPage = (props) => (
    <button onClick={props.startLogin}>Login</button>
);

const mapDispatchToProps = (dispatch) => ({
    startLogin: () => dispatch(startLogin())
});

export default connect(undefined, mapDispatchToProps)(LoginPage);