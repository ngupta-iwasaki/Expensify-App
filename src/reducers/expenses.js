const expensesReducerDefaultState = [];

const expensesReducer = (state = expensesReducerDefaultState, action) => {
    switch(action.type) {
        case "ADD_EXPENSE":
            return [
                ...state,
                action.expense
            ];
        case "REMOVE_EXPENSE":
            return state.filter(({id}) => id !== action.id);
        case "EDIT_EXPENSE":
            return state.map(exp => 
                exp.id !== action.id ? exp : {...exp, ...action.updates});
        case "SET_EXPENSES":
                return action.expenses;
        default:
            return state;
    }
};

export default expensesReducer;