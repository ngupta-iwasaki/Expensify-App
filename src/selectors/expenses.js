import moment from "moment";

const getVisibleExpenses = (expenses, {text, sortBy, startDate, endDate}) => {
    return expenses.filter(exp => {
        const createdAtMoment = moment(exp.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, "day") : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, "day") : true;
        const textMatch = exp.description.toLowerCase().includes(text.toLowerCase());

        return startDateMatch && endDateMatch && textMatch;
    }).sort((exp1, exp2) =>
            sortBy === "amount" ? exp2.amount - exp1.amount: 
                                    exp2.createdAt - exp1.createdAt);
};

export default getVisibleExpenses;