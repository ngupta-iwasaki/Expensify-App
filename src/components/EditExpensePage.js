import React from 'react';
import ExpenseForm from "./ExpenseForm";
import {startEditExpense, startRemoveExpense} from "../actions/expenses";
import {connect} from "react-redux";

export class EditExpensePage extends React.Component {
    
    onSubmit = (updates) => {
        this.props.startEditExpense(this.props.expense.id, updates);
        this.props.history.push("/");
    };

    onClick = () => {
        this.props.startRemoveExpense({id: this.props.expense.id});
        this.props.history.push("/");
    };

    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Edit Expense</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm 
                        expense={this.props.expense}
                        onSubmit={this.onSubmit}/>
                    <button className="button button--secondary"
                            onClick={this.onClick}>
                        Remove Expense</button>
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state, props) => ({
    expense: state.expenses.find((exp) => 
        exp.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch) => ({
    startRemoveExpense: (id) => dispatch(startRemoveExpense(id)),
    startEditExpense: (id, updates) => dispatch(startEditExpense(id, updates))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);