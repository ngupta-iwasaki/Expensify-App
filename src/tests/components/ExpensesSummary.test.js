import React from "react";
import {shallow} from "enzyme";
import {ExpensesSummary} from "../../components/ExpensesSummary";
import expenses from "../fixtures/expenses";

test("should render ExpensesSummary with 1 expense", () => {
    const wrapper = shallow(<ExpensesSummary expenseCount={1}
                                expensesTotal={1294}/>);
expect(wrapper).toMatchSnapshot();
});

test("should render ExpensesSummary with multiple expenses", () => {
    const wrapper = shallow(<ExpensesSummary expenseCount={2}
                                            expensesTotal={3000}/>);
    expect(wrapper).toMatchSnapshot();
});