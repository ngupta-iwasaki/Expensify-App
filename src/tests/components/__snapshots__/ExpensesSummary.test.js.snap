// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary with 1 expense 1`] = `
<h1>
  Viewing 
  1
   expense 
  totalling 
  $12.94
</h1>
`;

exports[`should render ExpensesSummary with multiple expenses 1`] = `
<h1>
  Viewing 
  2
   expenses 
  totalling 
  $30.00
</h1>
`;
